[MP Rewrite Part 4] Motion Planning Primitives

kylestach opened the pull request:
Primitives for motion planning including angle planning, RRT, velocity profiling, Bezier fitting, and replanning. This is anticipated to be the largest portion of the motion planning rewrite.
    
This is not expected to pass CI.

kylestach committed:
[MP Rewrite Part 4] Motion Planning Primitives

Primitives for motion planning including angle planning, RRT, velocity
profiling, Bezier fitting, and replanning. This is anticipated to be the
largest portion of the motion planning rewrite.

This is not expected to pass CI.

kylestach committed:
Merge branch 'mp-rewrite/trajectory' into mp-rewrite/primitives

ghost commented:
I love how trajectories might actually use a reasonable number of points now xD

oswinso commented on soccer/planning/low_level/AnglePlanning.hpp:
```diff
@@ -0,0 +1,83 @@
+
+namespace AngleFns {
+
+inline double tangent(const LinearMotionInstant& instant, double previous_angle,
+                      Eigen::Vector2d* jacobian) {
```
doxygen docstring?

oswinso commented on soccer/planning/low_level/CreatePath.hpp:
```diff
@@ -0,0 +1,27 @@
+namespace Planning::CreatePath {
+
+/**
+ * Generate a smooth path from start to goal avoiding obstacles.
```
brief

oswinso commented on soccer/planning/low_level/PathSmoothing.hpp:
```diff
@@ -0,0 +1,63 @@
+class BezierPath {
+public:
+    /**
+     * Construct a bezier path for a curve with a given velocity, going through
+     * the specified points and with specified velocity at endpoints.
```
brief

oswinso commented on soccer/planning/low_level/Replanner.hpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace Planning {
+
+class Replanner {
```
docstrings for this class?

oswinso commented on soccer/planning/low_level/TrapezoidalMotion.hpp:
```diff
@@ -0,0 +1,24 @@
+
+namespace Planning::Trapezoid {
+
+struct State {
+    double position;
+    double velocity;
+};
+
+double timeRemaining(State initial, State goal, double max_velocity,
+                     double max_acceleration);
+
+State predictIn(State initial, State goal, double max_velocity,
+                double max_acceleration, double time_now);
+
+std::optional<State> predictWithExactEndTime(State initial, State goal,
```
docstrings pls

oswinso commented on soccer/planning/low_level/VelocityProfiling.hpp:
```diff
@@ -0,0 +1,23 @@
+namespace Planning {
+
+/**
+ * Create a trajectory with the given path by calculating the maximum possible
+ * velocity at each points given constraints on speed and acceleration.
```
brief

oswinso commented on soccer/planning/low_level/AnglePlanning.cpp:
```diff
@@ -0,0 +1,141 @@
+    RJ::Time start_time = start_instant.stamp;
+
+    // Clip the front of the trajectory.
+    // TODO(Kyle): Handle this slightly more gracefully, by only profiling the
+    //  required portion of the trajectory.
```
issue num

oswinso commented on soccer/planning/low_level/AnglePlanning.cpp:
```diff
@@ -0,0 +1,141 @@
+void PlanAngles(Trajectory* trajectory, const RobotInstant& start_instant,
+                const AngleFunction& angle_function,
+                const RotationConstraints& /* constraints */) {
+    RJ::Time start_time = start_instant.stamp;
```
`const`

oswinso commented on soccer/planning/low_level/AnglePlanning.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    // TODO(Kyle): If we are unable to finish the motion in time at all,
+    //  profile all the way until the end, and then append points to the
+    //  trajectory containing the remainder of the trapezoid motion.
+    for (int i = static_cast<int>(velocity.size() - 1); i > 0; i--) {
+        double next_heading = target_angles.at(i);
+        double next_velocity = velocity.at(i);
+
+        double target_heading = target_angles.at(i - 1);
+        double delta = fixAngleRadians(next_heading - target_heading);
+
+        RJ::Time time_next = trajectory->instant_at(i).stamp;
+        RJ::Time time_prev = trajectory->instant_at(i - 1).stamp;
+        double dt = RJ::Seconds(time_next - time_prev).count();
+
+        // Fit a quadratic (constant acceleration) with the given final
+        // velocity and final position, and attempt to hit the target
+        // heading. If acceleration is too high, clamp it and adjust the
+        // heading accordingly.
+        // d = (vi+vf)t/2 = vf*t - 1/2at^2
+        double acceleration =
+            2 * (next_velocity * dt - delta) / std::pow(dt, 2);
+
+        // Clamp acceleration to our given limits.
+        // TODO(Kyle): Also clamp based on maximum velocity. That shouldn't
+        //  be necessary in most situations though, as angle functions are
+        //  not expected to have rapid discontinuities (possible exception:
+        //  face point, when moving near the point).
+        acceleration = std::clamp(acceleration,
+                                  -constraints.maxAccel,
+                                  constraints.maxAccel);
+
+        double prev_velocity = next_velocity - acceleration * dt;
+        double prev_heading =
+            next_heading - dt * (next_velocity + prev_velocity) / 2;
+
+        target_angles.at(i - 1) = prev_heading;
+        velocity.at(i - 1) = prev_velocity;
+    }
+     */
```
why is this section commented out?

oswinso reviewed (changes requested):
docstrings and issue nums

kylestach committed:
Remove disabled for-loop in Processor

kylestach committed:
Merge branch 'mp-rewrite/plan-request' into mp-rewrite/primitives

kylestach committed:
Merge branch 'mp-rewrite/plan-request' into mp-rewrite/primitives

kylestach committed:
Fix TODOs and docs in planning primitives

JNeiger commented on soccer/planning/low_level/AnglePlanning.cpp:
```diff
@@ -0,0 +1,140 @@
+    // required portion of the trajectory.
+    *trajectory = trajectory->subTrajectory(start_time, trajectory->end_time());
+
+    if (trajectory->num_instants() < 2) {
```
Is it possible to have 0 instants?

kylestach replied:
If it's empty it shouldn't be profiled (empty is used to represent planning failure).

JNeiger commented on soccer/planning/low_level/AnglePlanning.cpp:
```diff
@@ -0,0 +1,140 @@
+            Geometry2d::Point(gradient));
+    }
+
+    /*
```
Large commented out sections?

kylestach replied:
These are for #1506. Essentially, I'm just making the target angle independent of the current angle - in reality, we should replan using a trapezoid profile from the current state until we coincide with the target angle trajectory.

JNeiger commented on soccer/planning/low_level/AnglePlanning.hpp:
```diff
@@ -0,0 +1,103 @@
+ * used to decide whether to target forwards or backwards motion.
+ */
+inline double tangent(const LinearMotionInstant& instant, double previous_angle,
+                      Eigen::Vector2d* jacobian) {
```
Is this used anywhere?

kylestach replied:
Yes. Path target planner uses it.

JNeiger commented on soccer/planning/low_level/PathSmoothing.cpp:
```diff
@@ -0,0 +1,242 @@
+        //         ***
+        //            ***
+        //           c1  ****** p1 ---------> v1
+        // (* represents path, pi represents poi3nts, ci represents control
```
```suggestion
        // (* represents path, pi represents points, ci represents control
```

JNeiger commented on soccer/planning/low_level/PathSmoothing.cpp:
```diff
@@ -0,0 +1,242 @@
+#include "PathSmoothing.hpp"
```
There's a lot of math here, I assume this comes from a website or something, could you link it?

kylestach replied:
Nope. This was developed by reverse-engineering what the old code was doing, no idea where they got it from. I have a plan for replacing it but it's going to have to happen after this PR is all merged in.

JNeiger commented on soccer/planning/low_level/RRTUtil.cpp:
```diff
@@ -0,0 +1,133 @@
+        // probably doesn't exist
+        // a straight path.
+        biRRT.setMinIterations(0);
+        biRRT.setMaxIterations(5);
```
Why 5 times? Shouldn't it work the first time?

kylestach replied:
No, it has a maximum travel distance.

This should probably be replaced with just a straight-line hit check, though.

kylestach replied:
#1511

JNeiger replied:
I think it actually might since the step size is the distance between start and goal. I think we should still just swap to a straight line check since it's much more obvious what is happening

kylestach replied:
Yeah, you right. Super weird - anyway, I'll leave it as a TODO.

JNeiger commented on soccer/planning/low_level/RoboCupStateSpace.hpp:
```diff
@@ -0,0 +1,68 @@
+#pragma once
```
Was this changed?

kylestach replied:
No

JNeiger commented on soccer/planning/tests/RRTTest.cpp:
```diff
@@ -0,0 +1,306 @@
+
+#include "planning/low_level/RRTUtil.hpp"
+
+#if 0
```
Why is this disabled?

kylestach replied:
Mostly because I hadn't gotten around to re-enabling it yet. I'm not sure what the easiest way to do that with this sketchy merge strategy is, though - I may just add a pass re-enabling tests after parts 1-5 get merged into the base.

JNeiger reviewed (commented):
I'll have to look at this in more detail later. Just a quick overview for right now

kylestach committed:
Merge branch 'mp-rewrite/plan-request' into mp-rewrite/primitives

kylestach committed:
Address review comments

kylestach committed:
Merge branch 'mp-rewrite/trajectory' into mp-rewrite/primitives

oswinso reviewed (approved):
lgtm

JNeiger commented on soccer/planning/low_level/Replanner.cpp:
```diff
@@ -0,0 +1,187 @@
+
+bool Replanner::veeredOffPath(const Trajectory& trajectory, RobotInstant actual,
+                              RJ::Time now) {
+    constexpr double kReplanThreshold = 0.5;
```
This should probably be a config variable

JNeiger reviewed (approved):
LGTM, the config isn't blocking

kylestach committed:
Make off-path threshold configurable

kylestach committed:
[MP Rewrite Part 5] Port previous planners to new framework (#1497)

* [MP Rewrite Part 5] Port previous planners to new framework

Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

* Address review comments

* Move applyLowPassFilter into common, add TODOs

* Enable escape-obstacles planner

[MP Rewrite Part 6] Cleanup and fixes (#1512)

* Cleanup and fixes to make code run properly

* Fix test compilation and disable a trapezoidal motion profile edge case.

Some planner tests are still failing. This will be addressed in a later
commit.

* Prettify

* Make e2e planner tests pass.

Because they generate random scenarios, and the RRT planner is not
guaranteed to converge in a finite number of iterations (and the Bezier
approximation may be incorrect), these tests are currently flaky. That
should be fixed before this is merged to staging.

* Tidy fixes

* Linting

* Formatting

* Remove deleted file from CMakeLists, remove reuse condition from settle

* Don't replan in Collect::control

* Optimizations to geometry methods

* Reset and regenerate sim config

This deletes unnecessary variables and moves planning-related variables
into PathPlanner

* Formatting

* Add TODO in EscapeObstacles

* Tidy fixes

* Revert inconsistent clang-tidy recommended fix

* Temporarily disable flaky planner test

* Add NOLINTNEXTLINE

* Remove old sim config and add TODOs for trajectory utils

kylestach committed:
Merge branch 'mp-rewrite/plan-request' into mp-rewrite/primitives
